src/main/java/graphql/analysis/QueryTraversal.java

```diff
@@ -99,40 +97,36 @@ private GraphQLObjectType getRootType() {
         visitPreOrder((env) -> acc[0] = queryReducer.reduceField(env, (T) acc[0]));
         return (T) acc[0];
     }
-        
-    private List<Selection> childrenOf (Selection n) {
-        return selectionProvider.getChildren(n, fragmentsByName);
+
+    private List<Node> childrenOf(Node selection) {
+        return childrenOfSelectionProvider.getSelections((Selection) selection);
     }
-    
-    private void visitImpl(QueryVisitor visitor, SelectionSet selectionSet, GraphQLCompositeType type, QueryVisitorEnvironment parent, boolean preOrder) {
-        QueryTraversalContext context = new QueryTraversalContext(type, parent);
-        QueryTraversalDelegate delegate = preOrder
-                    ? new QueryTraversalDelegate(visitor::visitField, env -> {})
-                    : new QueryTraversalDelegate(env -> {}, visitor::visitField);
-        
-        // in order not to check parentContext for null and guarantee that
-        // we always can obtain the root QueryTraversalContext,
-        // we are subclassing here TraverserStack to set up a BARRIER
-        // parent that stores root QueryTraversalContext
-        Traverser.depthFirst(this::childrenOf)
-            .rootVar(QueryTraversalContext.class, context)
-            .traverse(selectionSet.getSelections(), null, delegate);
+
+    private void visitImpl(FieldVisitor visitFieldCallback, SelectionSet selectionSet, GraphQLCompositeType type, boolean preOrder) {
+        Map<Class<?>, Object> rootVars = new LinkedHashMap<>();
+        rootVars.put(QueryTraversalContext.class, new QueryTraversalContext(type, null));
+
+        NodeTraverser nodeTraverser = new NodeTraverser(rootVars, this::childrenOf);
+        nodeTraverser.depthFirst(new NodeVisitorImpl(visitFieldCallback, preOrder), selectionSet.getSelections());
     }
-    
-    private class QueryTraversalDelegate extends NodeVisitorStub<TraverserContext<Selection>> {
-        
-        final Consumer<QueryVisitorEnvironment> preOrder;
-        final Consumer<QueryVisitorEnvironment> postOrder;
-        
-        QueryTraversalDelegate (Consumer<QueryVisitorEnvironment> preOrder, Consumer<QueryVisitorEnvironment> postOrder) {
-            this.preOrder = assertNotNull(preOrder);
-            this.postOrder = assertNotNull(postOrder);
+
+    private class NodeVisitorImpl extends NodeVisitorStub {
+
+        private FieldVisitor visitFieldCallback;
+        boolean preOrder;
+
+        NodeVisitorImpl(FieldVisitor visitFieldCallback, boolean preOrder) {
+            this.visitFieldCallback = visitFieldCallback;
+            this.preOrder = preOrder;
         }
-        
+
         @Override
-        public Object visitInlineFragment(InlineFragment inlineFragment, TraverserContext<Selection> context) {
+        public TraversalControl visitInlineFragment(InlineFragment inlineFragment, TraverserContext<Node> context) {
+            if (context.getVar(LeaveOrEnter.class) == LEAVE) {
```

Comment:
Let's make this an explicit property of the `TraverserContext` in order to tell the Visitor the current traversal direction: down vs up ... or pre-order vs. post-order

Reply:
Perhaps code organization like
```java
switch (context.getVar(LeaveOrEnter.class)) {
   case LEAVE:
      return TraversalControl.CONTINUE;
   case ENTER:
      // do the stuff
}
```
will be more readable.


WDYT?